includes/class-sensei-frontend.php

```diff
@@ -744,39 +744,35 @@ public function sensei_complete_course() {
 			// Handle submit data.
 			switch ( $sanitized_submit ) {
 				case __( 'Mark as Complete', 'sensei-lms' ):
-					// Add user to course.
-					$course_metadata = array(
-						'start'    => current_time( 'mysql' ),
-						'percent'  => 0, // No completed lessons yet.
-						'complete' => 0,
-					);
-					$activity_logged = Sensei_Utils::update_course_status( $current_user->ID, $sanitized_course_id, 'in-progress', $course_metadata );
-
-					if ( $activity_logged ) {
-						// Get all course lessons.
-						$course_lesson_ids = Sensei()->course->course_lessons( $sanitized_course_id, 'any', 'ids' );
-						// Mark all quiz user meta lessons as complete.
-						foreach ( $course_lesson_ids as $lesson_item_id ) {
-							// Mark lesson as complete.
-							$activity_logged = Sensei_Utils::sensei_start_lesson( $lesson_item_id, $current_user->ID, true );
-						}
-
-						// Update with final stats.
-						$course_metadata = array(
-							'percent'  => 100,
-							'complete' => count( $course_lesson_ids ),
-						);
-						$activity_logged = Sensei_Utils::update_course_status( $current_user->ID, $sanitized_course_id, 'complete', $course_metadata );
-
-						do_action( 'sensei_user_course_end', $current_user->ID, $sanitized_course_id );
-
-						// Success message.
-						$this->messages = '<header class="archive-header"><div class="sensei-message tick">'
-							// translators: Placeholder is the Course title.
-							. sprintf( __( '%1$s marked as complete.', 'sensei-lms' ), get_the_title( $sanitized_course_id ) )
-							. '</div></header>';
+					$course_progress_repository = Sensei()->course_progress_repository_factory->create();
+					$course_progress            = $course_progress_repository->get( $sanitized_course_id, $current_user->ID );
```

Comment:
One aspect I wanted to bring up connected to this project is setting the foundations for previewing courses as a student, and even taking courses as guests. This would mean that queries and changes to student progress are not really for the current logged-in user., but a virtual one. I see two approaches to do this:

- Decouple relying on `$current_user` everywhere, and create a central method to get the current student. The preview functions can then filter this, return a virtual user and build new progress entries for them.

- Filter the progress repositories (or factories) directly, and change the state they return (or create new preview repositories with different implementation)

I'm thinking the first one is both easier and less prone to issues, it would allow the teacher to fully test a course with progress for the virtual student tracked across pages like it would be for a real one. 

Most of this is outside of the scope of the student progress project, but since we are replacing queries and updates to progress everywhere, it would be great to keep in sight and set the groundwork for.

Reply:
Good info, thanks! I like both options. Fetching the student from a dedicated method sounds great, also we might have to add some filters to allow for us or third parties to modify the progress data as well. But I think this is outside of the scope for now. Do you see any architectural changes that we need to do to account for this?

Reply:
I was thinking a bit more on this, and the best option might be to have a new repository that stores progress data in `$_SESSION`. Architecture-wise, looks like the one proposed here will be able to support that.

Reply:
Just commenting because I was studying a little bit about sessions and WordPress while working on #5569 (related thread here: p1662047207045139-slack-C8QJAM52B) . Looks like the best approach is to have a custom session handler, like WooCommerce has: https://github.com/woocommerce/woocommerce/blob/b82638c131b1d1f6c2f410871ca9f73905391621/plugins/woocommerce/includes/class-wc-session-handler.php

That way, we can avoid using PHP Sessions altogether, and it fits pretty well with the idea of having custom database tables for Sensei, so there's that. :) 

With that said, I agree that decoupling relying on `$current_user` looks like the best option on this case.

Reply:
Sounds good!

Just to summarize, this PR won't be updating the `$current_user` code because that's not part of the progress data layer.